elasticdl/pkg/ps/server.go

```diff
@@ -7,47 +7,82 @@ import (
 	"google.golang.org/grpc"
 	"log"
 	"net"
+	"sync"
 )
 
-type psServer struct {
+// Server defines servicer of ps
+type Server struct {
 	pb.PserverServer
+	Param *Parameter
+	Opt   Optimizer
+	PSID  int32
+	lock  *sync.Mutex
 }
 
-func (s *psServer) PullVariable(ctx context.Context, in *pb.PullVariableRequest) (*pb.PullVariableResponse, error) {
+// NewServer creates a Server instance
+func NewServer(PSID int32, opt string, lr float32) *Server {
+	var ps Server
+	ps.Param = NewParameter()
+	if opt == "SGD" {
+		ps.Opt = NewSGDOptimizer(lr)
+	}
+	ps.PSID = PSID
+	ps.lock = &sync.Mutex{}
+	return &ps
```

Comment:
You cannot return the address of a variable ps created in the stack.  Once the function call completes and the stack frame pops, the address becomes invalid.

The correct way is 

```go
return &Server {
    Param : NewParameter(),
    Opt: NewOptimizer(opt),
    ID: ID
}
```

It doesn't make sense to write

```go
ps.lock = & sync.Mutex()
```

because `lock ` should be defined in type Server as a variable but not an address:

```go
type Server {
    lock sync.Mutex
}
```

Reply:
1. Go support escape analysis. Please refer to https://stackoverflow.com/questions/38234487/returning-a-pointer-on-stack . And there is also a more detailed article https://www.ardanlabs.com/blog/2017/05/language-mechanics-on-escape-analysis.html.

When return a pointer inside a function, it escapes from the stack to heap automatically. So it's fine to define a server variable inside a function, and return its address.

2. Yes, I think `sync.Mutex` is more clear, what about following code:

```go
type Server struct {
	pb.PserverServer
	Param *Parameter
	Opt   Optimizer
	ID  int
	lock  sync.Mutex
}

func NewOptimizer(opt string, lr float32) Optimizer {
    if opt == "SGD" {
        return NewSGDOptimizer(lr)
    }
    return nil
}

// NewServer creates a Server instance
func NewServer(ID int, opt string, lr float32) *Server {
	return &Server{
        Param : NewParameter(),
        Opt: NewOptimizer(opt, lr),
        ID: ID}
}
```

Reply:
Parameter => Parameters, Model ?

Reply:
Model is more approriate. 
I am writing the concepts design doc. And it seems that the code change will be a little large. I will refactor the codes after the doc is done.